integration/container/stop_linux_test.go

```diff
@@ -64,3 +67,66 @@ func TestStopContainerWithTimeout(t *testing.T) {
 		})
 	}
 }
+
+// TestStopContainerWithTimeoutCancel checks that ContainerStop is not cancelled
+// if the request is cancelled or the connection is closed.
+// See issue https://github.com/moby/moby/issues/45731
+func TestStopContainerWithTimeoutCancel(t *testing.T) {
+	t.Parallel()
+	defer setupTest(t)()
+
+	const (
+		cancelContext   = "cancel context"
+		closeConnection = "close connection"
+	)
+
+	for _, tc := range []string{cancelContext, closeConnection} {
+		tc := tc
+		t.Run(tc, func(t *testing.T) {
+			t.Parallel()
+			apiClient := testEnv.APIClient()
+			defer apiClient.Close()
+			ctx := context.Background()
+
+			testCmd := container.WithCmd("sh", "-c", "sleep 30")
+			id := container.Run(ctx, t, apiClient, testCmd)
+
+			ctxCancel, cancel := context.WithCancel(context.Background())
+			stopTimeout := 3
+			defer cancel()
+			go func() {
+				err := apiClient.ContainerStop(ctxCancel, id, containertypes.StopOptions{Timeout: &stopTimeout})
+				switch tc {
+				case cancelContext:
+					assert.Check(t, is.ErrorType(err, errdefs.IsCancelled))
+				case closeConnection:
+					assert.Check(t, err)
+				}
```

Comment:
There's no synchronization between this goroutine and the test. These checks could potentially run after the test function has returned. I recommend sending the error over a channel and performing these checks on the test goroutine, e.g.:
```go
stoppedCh := make(chan error)
go func() {
    stoppedCh <- apiClient.ContainerStop(ctxCancel, ...)
}()

// [...]

var stoppedErr error
select {
case stoppedErr = <-stoppedCh:
case <-time.After(...):
    t.Fatal("Didn't stop??")
}
// Assert stoppedErr
```

Reply:
Ah, thanks; yes, that probably works